---
title: "A brief introduction to BamLiner"
author:
- name: "Roberto Amarie"
  affiliation: UniversitÃ  degli studi di Milano & Politecnico di Milano, Italy
  email: roberto.amarie@mail.polimi.it    or    roberto.amarie@studenti.unimi.it  
date: "`r Sys.Date()`"    
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{BamLiner}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
---


```{r, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
```

```{r style, echo = FALSE, results = 'asis'}
library(BiocStyle)
```

```{r, echo = FALSE}
library(knitr)
```

# Introduction

The R package **BamLiner** provides a function that extracts the primary
alignment of a desired read and enables the visualization of its
alignment along its reference sequence, as an useful useful quality
control inspection tool. As to achieve this, the package also includes
some internal functions for managing CIGAR operations, determining the
reference sequence of the full read, as well as the recomputing the
consumed read and reference sequences upon accounting for the CIGAR
operations.

------------------------------------------------------------------------

# Installing and loading the package {#installation}

## Installation

### Bioconductor

**BamLiner** requires several CRAN and Bioconductor R packages to be
installed. Dependencies are usually handled automatically, when
installing the package using the following commands:

```         
install.packages("BiocManager")
BiocManager::install("BamLiner")
```

[NOTE: Ignore the first line if you already have installed the
`r CRANpkg("BiocManager")`.]

### Manual installation

In the unlikely case that a manual installation is required, e.g., if
you do not install **BamLiner** via Bioconductor (which is highly
recommended), before installing **BamLiner** make sure the following
packages are installed:

Bioconductor packages:

`r Biocpkg("GenomicRanges")`, `r Biocpkg("Biostrings")`,
`r Biocpkg("IRanges")`, `r Biocpkg("S4Vectors")`,
`r Biocpkg("BSgenome.Hsapiens.UCSC.hg38")`,`r Biocpkg("BSgenome.Hsapiens.UCSC.hg38")`

Bioconductor packages can be installed from R using the following
command:

```         
BiocManager::install("<package_name>")
```

Sometimes, it may also be useful to update Bioconductor:

```         
BiocManager::install()
```

Finally, the manual installation of **BamLiner** can, for example, be
done from the command line ...

```         
R CMD INSTALL BamLiner_<version>.tar.gz
```

... or the newest version can directly be installed from GitHub using
the CRAN package `r CRANpkg("devtools")`:

```         
library(devtools)
install_github("robertoamarie/BamLiner")       #######well maybe I have to remove this part if I don't get to push it on my github
```

## Loading the package

After installation, loading the package is simple:

```{r, results='hide', message=FALSE}
library(BamLiner)
```

------------------------------------------------------------------------

# Input data {#input}

**BamLiner** works with SAM/BAM files resulting from any genome
sequencing experiment, except for those including SA:Z tagged chimeric
reads that will be accounted for in future versions of the tool. For
detailed information on this format (including an example), see
<https://samtools.github.io/hts-specs/SAMv1.pdf>.

Specifically, it requires that the files should be pre-loaded in R in
the form of a `r Rpackage("GAlignments")` object that includes sequence,
flag and qname information as metadata columns.

Additionally, **BamLiner** requires the genomic sequence (in form of a
`r Rpackage("BSgenome")` object) to determine the reference sequences on
which reads should map to.

------------------------------------------------------------------------

# Workflow {#workflow}

Once both the reference genome and the bam file have been loaded in the
environment in the required format specified in the above section, the
alignment of a given read to its reference sequence can be simply
determined using the *extract_read_by_qname_and_show_alignment*
function. The user can also choose if the softclipped portions of the
read will be highlighted by a red coloring of the print. Additionally, a
data.frame containing all the information needed for the visualization
can be outputted rather than the printed outcome.

**Example workflows:**

As a simple example, two subsampled bam files have been provided in the
package, one representing alignments from a BulkRNA-seq experiment while
the other representing a subsample of one of the processed outputs of a
PARIS experiment, chosen given its presence of gapped reads with small
"N" CIGAR operations. (data generated internally to the lab, full
availability not yet granted; to have at least some partial background
on the technique please refer to Zhang et al. Genome Research 2022
<https://doi.org/10.1101/gr.275979.121> & Lu et al. Cell 2016
<https://doi.org/10.1016/j.cell.2016.04.028>.)

```{r}
# load the reference genome
library(BSgenome.Hsapiens.UCSC.hg38)
ref_genome <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38


library(GenomicAlignments)
# load example bam represnting a subset of a human brain Bulk RNA-seq experiment (GEO:GSE219444)
# downloaded from ENCODE (https://www.encodeproject.org/files/ENCFF047UEJ/@@download/ENCFF047UEJ.bam) then subsetted via samtools view



sys_bam <- system.file("extdata", "10k_RNAseq_subset.bam", package = "BamLiner")

bam <- readGAlignments(sys_bam,
                       param = ScanBamParam(what = c("seq","flag","qname")))



# Showing the alignment of an arbitrarily chosen read
read <- mcols(bam)$qname[1555]  
extract_read_by_qname_and_show_alignment(bam, read, ref_genome)


# M containing read 
read <- mcols(bam)$qname[18]  
extract_read_by_qname_and_show_alignment(bam, read, ref_genome)

# S
read <- mcols(bam)$qname[1]  
extract_read_by_qname_and_show_alignment(bam, read, ref_genome)

# D
read <- mcols(bam)$qname[287]  
extract_read_by_qname_and_show_alignment(bam, read, ref_genome)

# I
read <- mcols(bam)$qname[485]  
extract_read_by_qname_and_show_alignment(bam, read, ref_genome)


## example index values have been identified by means of: 
# head(grep("D", bam@cigar))
#
#
# # Showing the alignment of the first read that contains a Deletion
# del_read_index <- head(grep("D", bam@cigar), n = 1)
# read <- mcols(bam)$qname[del_read_index]  
# extract_read_by_qname_and_show_alignment(bam, read, ref_genome)
# 
# 
# # Showing the alignment of the first read that contains an Insertion
# del_read_index <- head(grep("I", bam@cigar), n = 1)
# read <- mcols(bam)$qname[del_read_index]  
# extract_read_by_qname_and_show_alignment(bam, read, ref_genome)
```

```{r}
# load the reference genome
library(BSgenome.Hsapiens.UCSC.hg38)
ref_genome <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38


# load example bam representing a subset of iPSC PARIS data generated internally by my lab at HT
# processed as to contain only the primary alignments of non-chimeric yet gapped reads
sys_gap1 <- system.file("extdata", "sampled_monoaligned.bam", package="BamLiner")

gap1 <- readGAlignments(sys_gap1,
                       param = ScanBamParam(what = c("seq","flag","qname")))



# N containg read
read <- mcols(gap1)$qname[17]
extract_read_by_qname_and_show_alignment(gap1, read, ref_genome)




# N & highlighted S
read <- mcols(gap1)$qname[1]
extract_read_by_qname_and_show_alignment(gap1, read, ref_genome)

# without soft-clip highlighting
read <- mcols(gap1)$qname[303]
extract_read_by_qname_and_show_alignment(gap1, read, ref_genome, colored_softclips = FALSE)




# outputing the list rather than the print
read <- mcols(gap1)$qname[303]
extract_read_by_qname_and_show_alignment(gap1, read, ref_genome, print_output = FALSE)

```

------------------------------------------------------------------------

# Internal functions {#internal_functions}

IDK mabye add a section like this?

Despite treated as internal functions for the scope of this package as
to simplify the end-user experience in utilizing them as intended, the
additional functions developed represent useful utils that can be still
used by advanced users seeking more flexibility.

```{r, eval= FALSE}
BamLiner:::.parseCIGAR()

BamLiner:::.fetch_ref_seq()

BamLiner:::.show_alignment_to_reference()
```

```{r}
BamLiner:::.parseCIGAR("2S4M1I7M2D5H")
```

```{r}
read <- mcols(bam)$qname[1555] 
read_granges <- granges(bam[1555])
cig_ops <-BamLiner:::.parseCIGAR("2S4M1I7M2D5H")


BamLiner:::.fetch_ref_seq(ref_genome,
                          read_granges,
                          cig_ops)
```

------------------------------------------------------------------------

# Session Info {#session-info}

The following R session information was used to generate this vignette:

```{r session-info}
sessionInfo()
```
